### PR TITLE
Add 'warning density' computation to the warnhist script

### PR DESCRIPTION
**Pull request by qkoziol**
Add 'warning density' computation to the warnhist script, along with several cleanups to it.   Add "--enable-show-all-warnings" configure (and CMake) option to disable compiler diagnostic suppression (and therefore show all the
otherwise suppressed compiler diagnostics), disabled by default.  Clean up a buncn of misc. warnings.

### CODE REVIEW

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:103]: Warning density computation requires knowing where the source code files are located (so that their lines of code can be computed, for the density calculation).  So, I added the '-p <path list>' option as well.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:113]: I also added a '-q' ("quiet") option to suppress warnings from the script itself.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:89]: Reflowed lines to display better, no text change.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:92]: Reflowed lines to display better, no text change.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:75]: Added new '-d', '-p', and '-q' options to command-line summary.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:62]: Used for warning density calculation

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:25]: Used for warning density calculation

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:119]: Count lines in a [source] file.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:136]: Store the paths to source file names in a directory hierarchy, for later use.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:182]: Locate source files, counting their lines of code.  Handles multiple source files correctly, as long as the number of compiles equals the number of times a file of that name is found in the directory hierachy(s).

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:231]: Determine the type of source file being compiled, then locate it in the source directory hierarchy(s), and count its lines of code.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:254]: Make certain that the number of compiles of a source file name equals the number of files with that name.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:278]: Add new command-line options.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:374]: Scan the source file paths if warning density is requested.  (Note that '.' is used if no source file paths are given)

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:400]: Detect FORTRAN source & include files better.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:415]: Count lines-of-code for compiled files (lines that start with CC ..., FC ..., etc)

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:429]: Don't count Makefile warnings as compile warnings

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:435]: Detect more variants of using 'warning' in a variable or function name, etc.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:451]: Some invocations of the compiler now say '\<command-line\> instead of 'cc1'.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:241]: This was from an undetected FORTRAN include file and wasn't actually a CMAKE build warning, so detect FORTRAN include files (above) and remove this check.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:506]: Use Perl 'defined', instead of 'length' idiom.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:509]: Use Perl 'defined', instead of 'length' idiom.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:305]: Moved down, to avoid 'undefined variable' warning from Perl about $warning variable.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:594]: Add '%p' printf format.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:608]: New warning to genericize.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:665]: Try to make certain that the # of compiles of source files with same name matches the number of times files with that name are found in the source directory hierarchy(s).

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:684]: Add warning density output.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:724]: Fix: All warnings should be expanded with '-W'.

File: bin/warnhist
Comment (qkoziol) [inline bin/warnhist:762]: Fix: All files should be expanded with '-F'.

File: config/cmake/ConfigureChecks.cmake
Comment (qkoziol) [inline config/cmake/ConfigureChecks.cmake:597]: Add new CMake option to disable diagnostic suppression.

File: configure.ac
Comment (qkoziol) [inline configure.ac:2314]: Add new autotools option to disable diagnostic suppression.

File: fortran/src/H5config_f.inc.cmake
Comment (qkoziol) [inline fortran/src/H5config_f.inc.cmake:31]: Fix typo that was generating warning.

File: src/H5Dpkg.h
Comment (qkoziol) [inline src/H5Dpkg.h:730]: Add attribute to address warning.

File: src/H5EApkg.h
Comment (qkoziol) [inline src/H5EApkg.h:392]: Add attribute to address warning.

File: src/H5ESpkg.h
Comment (qkoziol) [inline src/H5ESpkg.h:78]: Add attribute to address warning.

File: src/H5Ipkg.h
Comment (qkoziol) [inline src/H5Ipkg.h:62]: Allow an ID's object to be treated as 'const' or not 'const' as needed, avoiding warnings / diagnostic suppression noise in H5I code.

File: src/H5Pprivate.h
Comment (qkoziol) [inline src/H5Pprivate.h:178]: Add attribute to address warning.

File: src/H5Tpkg.h
Comment (qkoziol) [inline src/H5Tpkg.h:873]: Add attribute to address warning.

File: test/event_set.c
Comment (qkoziol) [inline test/event_set.c:26]: Add attribute to address warning.

File: src/H5private.h
Comment (qkoziol) [inline src/H5private.h:500]: When the 'show all warnings' build option is used, don't suppress compiler diagnostics.

File: tools/src/h5import/h5import.c
Comment (qkoziol) [inline tools/src/h5import/h5import.c]: Add attribute to address warning.

File: utils/mirror_vfd/mirror_writer.c
Comment (qkoziol) [inline utils/mirror_vfd/mirror_writer.c:32]: Unused.

**Review by lrknox — APPROVED** [review]

**Review by bmribler — APPROVED** [review]

**Comment by lrknox** [comment]
@derobins, @brtnfld, @epourmal, this PR needs a review from one of you as Fortran code owner. 
Never mind.  The only Fortran change was to correct a CMake variable name.  I'm going to merge it.

**Pull request merged**
